bpaotu/bpaotu/views.py

```diff
@@ -495,6 +496,213 @@ def str_none_blank(v):
         return csv_fd.getvalue()
 
 
+@require_http_methods(["GET"])
+def otu_biom_export(request):
+    zf = zipstream.ZipFile(mode='w', compression=zipstream.ZIP_DEFLATED)
+    fd = StringIO()
+
+    def generate_biom_file():
+        '''
+        Write out the JSON file header first
+        '''
+        biom_file = OrderedDict()
+        biom_file['id'] = None
+        biom_file['format'] = "1.0.0"
+        biom_file['format_url'] = "http://biom-format.org"
+        biom_file['type'] = "OTU table"
+        biom_file['generated_by'] = "Bioplatforms Australia"
+        biom_file['date'] = datetime.datetime.now().replace(microsecond=0).isoformat()
+        biom_file['matrix_type'] = "sparse"
+        biom_file['matrix_element_type'] = "int"
+
+        biom_header = "{"
+
+        for key, val in biom_file.items():
+            biom_header += '"{}": "{}",\n'.format(key, val)
+
+        fd.write(biom_header)
+        yield fd.getvalue().encode('utf8')
+        fd.seek(0)
+        fd.truncate(0)
+
```

Comment:
you shouldn't be writing to the `fd` here; just yield, then do the writing in a `for` loop over the `generate_biom_file()` iterator. move `generate_biom_file` out; probably into a new `biom.py`. its purpose is then to make a BIOM file when passed a SampleQuery, yielding back the strings. if we want to stream a zip back, that's in the request handler, if we want to write to a file, we can do that in another function (e.g. for the celery stuff)

Reply:
Fixed.